azure-kusto-ingest/src/resource_manager/ingest_client_resources.rs

```diff
@@ -0,0 +1,169 @@
+    }
+
+    resource_uris.into_iter().collect()
+}
+
+/// Helper to turn a vector of resource URIs into a vector of Azure clients of type T with the provided [ClientOptions]
+fn create_clients_vec<T>(resource_uris: &[ResourceUri], client_options: &ClientOptions) -> Vec<T>
+where
+    T: ClientFromResourceUri,
+{
+    resource_uris
+        .iter()
+        .map(|uri| T::create_client(uri.clone(), client_options.clone()))
+        .collect()
+}
+
+/// Storage of the clients required for ingestion
+#[derive(Debug, Clone)]
+pub struct InnerIngestClientResources {
+    pub ingestion_queues: Vec<QueueClient>,
+    pub temp_storage_containers: Vec<ContainerClient>,
+}
+
+impl TryFrom<(&TableV1, &QueuedIngestClientOptions)> for InnerIngestClientResources {
+    type Error = IngestionResourceError;
+
+    /// Attempts to create a new InnerIngestClientResources from the given [TableV1] and [QueuedIngestClientOptions]
+    fn try_from(
+        (table, client_options): (&TableV1, &QueuedIngestClientOptions),
+    ) -> std::result::Result<Self, Self::Error> {
+        let secured_ready_for_aggregation_queues =
+            get_resource_by_name(table, "SecuredReadyForAggregationQueue".to_string())?;
+        let temp_storage = get_resource_by_name(table, "TempStorage".to_string())?;
+
+        Ok(Self {
+            ingestion_queues: create_clients_vec(
+                &secured_ready_for_aggregation_queues,
+                &client_options.queue_service,
+            ),
+            temp_storage_containers: create_clients_vec(
+                &temp_storage,
+                &client_options.blob_service,
+            ),
+        })
+    }
+}
+
+pub struct IngestClientResources {
+    client: KustoClient,
+    resources: ThreadSafeCachedValue<Option<InnerIngestClientResources>>,
+    client_options: QueuedIngestClientOptions,
+}
+
+impl IngestClientResources {
+    pub fn new(client: KustoClient, client_options: QueuedIngestClientOptions) -> Self {
+        Self {
+            client,
+            resources: Arc::new(RwLock::new(Cached::new(None, RESOURCE_REFRESH_PERIOD))),
+            client_options,
+        }
+    }
+
+    /// Executes a KQL management query that retrieves resource URIs for the various Azure resources used for ingestion
+    async fn query_ingestion_resources(&self) -> Result<InnerIngestClientResources> {
+        let results = self
+            .client
+            .execute_command("NetDefaultDB", ".get ingestion resources", None)
+            .await?;
+
+        let new_resources = results
+            .tables
+            .first()
+            .ok_or(IngestionResourceError::NoTablesFound)?;
+
+        InnerIngestClientResources::try_from((new_resources, &self.client_options))
+    }
+
+    /// Gets the latest resources either from cache, or fetching from Kusto and updating the cached resources
+    pub async fn get(&self) -> Result<InnerIngestClientResources> {
```

Comment:
This is similar to the one in ingest_client resources - can we export this to a trait maybe?

Reply:
Updated such that the logic is located in one place now